DM-39047: Switch from pkg_resources to importlib.resources

timj opened the pull request:
We could make further switches to `ResourcePath` for locating correction files but this solves the immediate problem of `pkg_resources` being deprecated.

timj committed:
Switch from pkg_resources to ResourcePath

We could make further switches to ResourcePath for locating
correction files but this solves the immediate problem of
pkg_resources being deprecated.

timj committed:
Fix type annotation

ktlim commented on python/astro_metadata_translator/headers.py:
```diff
@@ -354,17 +358,19 @@ def _find_from_resource(
     """
     if package is not None and resource_root is not None:
```
It seems strange that these are optional but nothing is done if they're not present.

timj replied:
it's because this is what `translator.resource_root` returns if there are no correction resources. Are you saying this should be the responsibility of the caller and this method should raise if it's not given a package and root?

ktlim commented on requirements.txt:
```diff
@@ -1,2 +1,3 @@
 astropy >=3.0.5
 pyyaml >=3.13
+lsst-resources
```
Shouldn't this go in the `ups/.table` file as well?

timj committed:
Use importlib.resources rather than ResourcePath

Defer using ResourcePath until we need the flexibility.

timj committed:
Drop python 3.8

The importlib.resources API only works with 3.9 and newer.
importlib_resources could be used if this is problematic.

timj committed:
Fix simple doc string issues

timj committed:
Fix docstrings for pydocstyle

timj committed:
Add ruff and pydocstyle configurations

timj committed:
Fix sphinx documentation problems

timj committed:
Add nitpick configuration

timj committed:
Add ruff linter

timj committed:
Turn on nitpick mode in sphinx build

timj committed:
Turn off pytest-flake8

timj committed:
Minor doc tweak

timj committed:
Only build the documentation once in the matrix

It fails on python 3.9 because of the type annotations.

timj committed:
Add explicit click dependency and remove click from types file

In click 8 the type annotations are already present.

ktlim commented on python/astro_metadata_translator/headers.py:
```diff
@@ -351,18 +351,19 @@ def _find_from_resource(
-                corrections = _read_yaml(fh, f"package resource {package}:{resource_name}")
+        resource_path = resources.files(package).joinpath(resource_root, target_file)  # type: ignore
+        if resource_path.is_file():
+            resource_uri = f"resource://{package}/{posixpath.join(resource_root, target_file)}"
```
Slightly worried that at some point we might get into trouble with things that are escaped differently in URIs vs. POSIX paths, but it's not worth doing anything about that now.

timj replied:
Right, if I used ResourcePath the escaping should be handled directly and the code would use `ResourcePath.join()`.
